### ISSUE TITLE
Cannot access secrets from container

### ISSUE THREAD

**Issue by amolenk** [Azure/service-fabric-mesh-preview#341]
I've added a secret to my ARM template like this:

```
{
      "apiVersion": "2018-09-01-preview",
      "name": "airportMeshStorageAccountKey",
      "type": "Microsoft.ServiceFabricMesh/secrets",
      "location": "[parameters('location')]",
      "dependsOn": [],
      "properties": {
        "kind": "inlinedValue",
        "contentType": "text/plain",
        "description": "Azure Storage Account key."
      }
    },
    {
      "apiVersion": "2018-09-01-preview",
      "name": "airportMeshStorageAccountKey/v1",
      "type": "Microsoft.ServiceFabricMesh/secrets/values",
      "location": "[parameters('location')]",
      "dependsOn": [
        "Microsoft.ServiceFabricMesh/secrets/airportMeshStorageAccountKey"
      ],
      "properties": {
        "value": "[parameters('storageAccountKey')]"
      }
    },
```

Now I'm trying to access that secret from inside my container. According to the docs, this should work:

```
var fabricSettingsPath = Environment.GetEnvironmentVariable("Fabric_SettingPath");
var secretPath = System.IO.Path.Combine(fabricSettingsPath, "airportMeshStorageAccountKey");
var connectionString = await File.ReadAllTextAsync(secretPath);
```

This doesn't work though because the `Fabric_SettingPath` is not set. I've printed all environment variables to the container log and it isn't there.

**Comment by aloneguid** [comment]
I can confirm that Fabric_Settings path is not set for me too.

**Comment by guibirow** [comment]
I am afraid it does not work!

Because Secrets are an independent resources, I don't think they will bind magically to all containers in the resource group. I assume you should have to link a secret to a service\code_package, but nowhere in the docs they reference how this should happen.


From what I could notice, the working way to pass settings to a containers without passing via environment variables is using the settings parameter in the code package. This works the way secrets should work([as described in the docs](https://docs.microsoft.com/en-us/azure/service-fabric-mesh/service-fabric-mesh-howto-manage-secrets#modify-mesh-application-to-reference-mesh-secret-values)) .

I am not sure if it is the way forward, but should be as simple like this:


    {
	"type": "Microsoft.ServiceFabricMesh/services",
	"name": "MyContainerService",
	"properties": {
	  "description": "My Service Fabric Mesh Service.",
	  "osType": "linux",
	  "codePackages": [
		{
		  "name": "MyContainer",
		  "image": "MyContainer",
		  "endpoints": [
			{
			  "name": "myListener",
			  "port": "80"
			}
		  ],
		  "resources": {
			"requests": {
			  "cpu": "1",
			  "memoryInGB": "1"
			}
		  },
		  "environmentVariables": [
			{
			  "name": "testing",
			  "value": "123"
			}
		  ],
		  "settings": [
			{
			  "name": "MySettings",
			  "value": "qwertyuiop123"
			}
		  ]
		}
	  ],
	  "replicaCount": "1",
	  "networkRefs": [
		{
		  "name": "[resourceId('Microsoft.ServiceFabricMesh/networks', 'helloWorldNetwork')]",
		  "endpointRefs": [
			{
			  "name": "myListener"
			}
		  ]
		}
	  ]
	}
    }

Each settings will be listed as a file in the settings folder, the location of the folder is defined in the `Fabric_SettingPath` Environment Variable`

On **linux**, you should be able to see the `MySettings` as a file in the path: `/var/settings/MySettings`

On **Windows**, you should be able to see the `MySettings` as a file in the path: `C:\Settings\MySettings`

**Comment by amolenk** [comment]
I can confirm that adding settings to the template does set the `Fabric_SettingPath` env variable, so that seems to work. I'm curious however how this could work together with secrets. Maybe something like:

```
"settings":
{
	"name": "MySettings",
	"value": "[resourceId('Microsoft.ServiceFabricMesh/secrets/values','mySecretSetting','v1')]"
}
```
I haven't had time to test this out (will do so tomorrow), but this same pattern is used when specifying the image registry password.

**Comment by aloneguid** [comment]
Is there any way to declare settings in yaml files as from VS solution?

**Comment by guibirow** [comment]
I haven't tried, but I would try like this:

    services:
      - name: helloWorldService
        properties:
          description: Hello world service.
          osType: linux
          codePackages:
            - name: helloworld
              image: myapp:1.0-alpine
              resources:
                requests:
                  cpu: 2
                  memoryInGB: 2
              endpoints:
                - name: helloWorldEndpoint
                  port: 8080
              settings:
                - name: MySettings
                  value: qwertyuiop123
    replicaCount: 3
    networkRefs:
      - name: mynetwork

Mainly, settings is an array attribute for CodePackage object, the above should work

**Comment by aloneguid** [comment]
@guibirow nah, that generates ARM without secrets resource and is completely ignored by Azure.

**Comment by guibirow** [comment]
Probably the VS extension needs to be updated then!
I haven't been using it, so can't confirm

**Comment by guibirow** [comment]
@mattrowmsft Do you have access to any example on how to link the Secret\Values to the containers?
The documentation does not show a valid example to link a existing secret to a container.

The approach I mentioned above(using settings) is not documented anywhere either.

**Comment by mattrowmsft** [comment]
@mostafaez do you have something shareable?

**Comment by sirdank** [comment]
There is a working sample for this here:

https://github.com/Azure-Samples/service-fabric-mesh/blob/master/templates/helloworld/helloworld.private_registry.windows.json

I'm going to open a ticket for updating the doc on using secrets.

**Comment by guibirow** [comment]
@sirdank If I understood well the sample provided, SF Mesh RP will load the secret at deployment time, and inject it into the ARM template. Is it correct?

In this case, if we inject it into the container, when the secret values change, do we need to  redeploy the code package to get new values?

**Comment by sirdank** [comment]
@guibirow 1. Yes, that part is working for me. I believe it would still be working if I just used an ordinary setting though.
2. I'm not sure but I believe that's what the doc is saying. This has led me to an additional question mentioned above. I'm unclear why we should use this feature if we have to reference the secret in our codePackage like this:

"settings": [
                    {
                      "name": "MySecretPasswordSetting",
                      "value": "[resourceId('Microsoft.ServiceFabricMesh/secrets/values','Certificate','1')]"
                    }
                  ]

It would be cleaner and equally secure just to add a securestring parameter and pass that directly to the settings node, as far as I can tell.

**Comment by guibirow** [comment]
Yes, what I wonder is why we need to hard bind the secret value to the container, in case it changes, it will require a new deployment.

I was expecting to have the secret (not the value) bonded to the container, whenever the value changes, it updates the container without needing a new deployment.

Let's wait for an official answer!
